File: sdks/python/settlekit/resources/usage.py
"""Usage resource — metering and prepaid credits."""

from __future__ import annotations

from typing import Any, Optional
from urllib.parse import quote

from ._base import AsyncResource, SyncResource, clean_body


def _meter_body(
    *,
    organization_id: str,
    customer_id: str,
    product_id: str,
    metric: str,
    quantity: int,
    period_start: Optional[str],
) -> dict[str, Any]:
    return clean_body(
        {
            "organizationId": organization_id,
            "customerId": customer_id,
            "productId": product_id,
            "metric": metric,
            "quantity": quantity,
            "periodStart": period_start,
        }
    )


def _credit_body(
    *, organization_id: str, customer_id: str, product_id: str, credits: int
) -> dict[str, Any]:
    return {
        "organizationId": organization_id,
        "customerId": customer_id,
        "productId": product_id,
        "credits": credits,
    }


def _credits_query(*, organization_id: str, customer_id: str, product_id: str) -> str:
    return (
        f"/v1/usage/credits?organizationId={quote(organization_id)}"
        f"&customerId={quote(customer_id)}&productId={quote(product_id)}"
    )


class Usage(SyncResource):
    """Synchronous usage-based billing namespace."""

    def record(
        self,
        *,
        organization_id: str,
        customer_id: str,
        product_id: str,
        metric: str,
        quantity: int = 1,
        period_start: Optional[str] = None,
    ) -> Any:
        """Record usage of a metric. ``POST /v1/usage/record``."""
        return self._request(
            "POST",
            "/v1/usage/record",
            body=_meter_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                metric=metric,
                quantity=quantity,
                period_start=period_start,
            ),
        )

    def credits(self, *, organization_id: str, customer_id: str, product_id: str) -> Any:
        """Read a prepaid balance. ``GET /v1/usage/credits``."""
        return self._request(
            "GET",
            _credits_query(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
            ),
        )

    def grant_credits(
        self, *, organization_id: str, customer_id: str, product_id: str, credits: int
    ) -> Any:
        """Grant prepaid credits. ``POST /v1/usage/credits/grant``."""
        return self._request(
            "POST",
            "/v1/usage/credits/grant",
            body=_credit_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                credits=credits,
            ),
        )

    def consume_credits(
        self, *, organization_id: str, customer_id: str, product_id: str, credits: int = 1
    ) -> Any:
        """Consume prepaid credits. ``POST /v1/usage/credits/consume``."""
        return self._request(
            "POST",
            "/v1/usage/credits/consume",
            body=_credit_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                credits=credits,
            ),
        )


class AsyncUsage(AsyncResource):
    """Asynchronous usage-based billing namespace."""

    async def record(
        self,
        *,
        organization_id: str,
        customer_id: str,
        product_id: str,
        metric: str,
        quantity: int = 1,
        period_start: Optional[str] = None,
    ) -> Any:
        """Record usage of a metric. ``POST /v1/usage/record``."""
        return await self._request(
            "POST",
            "/v1/usage/record",
            body=_meter_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                metric=metric,
                quantity=quantity,
                period_start=period_start,
            ),
        )

    async def credits(
        self, *, organization_id: str, customer_id: str, product_id: str
    ) -> Any:
        """Read a prepaid balance. ``GET /v1/usage/credits``."""
        return await self._request(
            "GET",
            _credits_query(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
            ),
        )

    async def grant_credits(
        self, *, organization_id: str, customer_id: str, product_id: str, credits: int
    ) -> Any:
        """Grant prepaid credits. ``POST /v1/usage/credits/grant``."""
        return await self._request(
            "POST",
            "/v1/usage/credits/grant",
            body=_credit_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                credits=credits,
            ),
        )

    async def consume_credits(
        self, *, organization_id: str, customer_id: str, product_id: str, credits: int = 1
    ) -> Any:
        """Consume prepaid credits. ``POST /v1/usage/credits/consume``."""
        return await self._request(
            "POST",
            "/v1/usage/credits/consume",
            body=_credit_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                credits=credits,
            ),
        )

File: sdks/rust/src/resources/usage.rs
//! Usage-based billing resource (`/v1/usage`) — metering + prepaid credits.

use reqwest::Method;
use serde::Serialize;

use crate::{Client, Result};

/// Accessor for usage-based billing operations.
#[derive(Debug)]
pub struct Usage<'a> {
    client: &'a Client,
}

/// Body for recording usage (`POST /v1/usage/record`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordUsage {
    pub organization_id: String,
    pub customer_id: String,
    pub product_id: String,
    pub metric: String,
    pub quantity: i64,
}

/// Body for granting/consuming prepaid credits.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreditOp {
    pub organization_id: String,
    pub customer_id: String,
    pub product_id: String,
    pub credits: i64,
}

impl<'a> Usage<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Usage { client }
    }

    /// Record usage of a metric (creates the meter on first use).
    pub async fn record(&self, body: &RecordUsage) -> Result<serde_json::Value> {
        self.client
            .request(Method::POST, "/v1/usage/record", Some(body))
            .await
    }

    /// Read a customer's prepaid credit balance for a product.
    pub async fn credits(
        &self,
        organization_id: &str,
        customer_id: &str,
        product_id: &str,
    ) -> Result<serde_json::Value> {
        let path = format!(
            "/v1/usage/credits?organizationId={}&customerId={}&productId={}",
            urlencoding(organization_id),
            urlencoding(customer_id),
            urlencoding(product_id),
        );
        self.client.request_no_body(Method::GET, &path).await
    }

    /// Grant prepaid credits.
    pub async fn grant_credits(&self, body: &CreditOp) -> Result<serde_json::Value> {
        self.client
            .request(Method::POST, "/v1/usage/credits/grant", Some(body))
            .await
    }

    /// Consume prepaid credits (meter a paid call).
    pub async fn consume_credits(&self, body: &CreditOp) -> Result<serde_json::Value> {
        self.client
            .request(Method::POST, "/v1/usage/credits/consume", Some(body))
            .await
    }
}

/// Minimal percent-encoding for query-string values (alphanumerics + a few safe
/// chars pass through; everything else is `%XX`).
fn urlencoding(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for byte in value.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(byte as char);
            }
            _ => out.push_str(&format!("%{byte:02X}")),
        }
    }
    out
}
